quil-rs/src/hash.rs

```diff
@@ -0,0 +1,78 @@
+use std::{
+    collections::{hash_map::DefaultHasher, HashMap},
+    hash::{Hash, Hasher},
+};
+
+/// Hashes a f64 using its u64 representation.
+///
+/// Notes:
+/// * This function hashes +0.0 and -0.0 to the same value.
+/// * The [documentation](https://doc.rust-lang.org/std/primitive.f64.html#method.to_bits) claims
+///   that this is generally portable in practice.
+#[inline]
+pub(crate) fn hash_f64<H: Hasher>(value: f64, state: &mut H) {
+    // +0.0 and -0.0 have different bits, so we use the bit form of +0.0 for both, that way
+    // hash(+0.0) == hash(-0.0)
+    if value == 0.0f64 {
+        0.0f64.to_bits().hash(state)
+    } else {
+        value.to_bits().hash(state)
+    }
+}
+
+/// Hashes a HashMap by taking the XOR sum of each key value pair.
+///
+/// This function uses XOR, because as a commutative operation, we get the same,
+/// unique XOR sum for each unique set of key value pairs, regardless of their order.
+#[inline]
+pub(crate) fn hash_hashmap<H: Hasher, K: Hash, V: Hash>(map: &HashMap<K, V>, state: &mut H) {
+    let mut final_hash = 0;
+    for element in map {
+        let mut hasher = DefaultHasher::new();
+        element.hash(&mut hasher);
+        final_hash ^= hasher.finish();
+    }
+
+    state.write_u64(final_hash)
+}
```

Comment:
This approach takes inspiration from how the rustc internals, and STLs in other languages hash unordered collections like `HashMap`s. Unfortunately, this algorithm conflicts with the default SipHash algorithm Rust currently uses for `HashMap`s, where the order of values is fundamentally important for achieving the security it offers. If we ignore the order with a commutative sum like this, it is theoretically possible to generate a set of values that will cause collisions and such a set could be used to attempt a DoS attack if this hashing algorithm is used somewhere inside of a service not expecting it. If hashing instructions is mostly a user utility then this seems like a non-issue, but as a library it isn't possible to discern how the feature will be used in all circumstances.

An alternative algorithm would be to take the hash of each key and value pair, sort them, then iterate over that list and use them to update the hasher state. This wouldn't betray the underlying SipHash algorithm, but it wouldn't have the expected runtime characteristics of a hashing function because of the sort and allocations needed.

We can also not allow any instructions containing a `HashMap` or similar to be hashed. This would impact `WaveformInvocation`s, `Capture`s, `Pulse`s, `FrameDefinition`s, and because they are all members of the `Instruction` enum, it also couldn't be hashed.

Reply:
I'm not really familiar with how hash-DoS attacks work or how SipHash helps protect against them, but why is the order important for the security guarantee?

Reply:
https://stackoverflow.com/questions/52184366/why-does-hashmap-need-a-cryptographically-secure-hashing-function

TL;DR specifically crafted input can cause lots of hash collisions, leading to your hash map runtime exploding and DOSing the service.

As for commutativity, if order isn’t important, then that makes it even easier to have hash collisionss.

Reply:
A couple of additional thoughts, after stewing and discussing with a few folks offline:

We're not sure how widely used this would be, and it may not be worth supporting this at all. There is an immediate use case for hashing instructions in pyQuil, but it may not need these specific instructions. In that case, we may omit support some instructions. If someone asks for it, we can take the opportunity to understand the use case and circle back. There are also other measure we could take to well to partially support hashing the `Instruction` type in Python if necessary.

Also, @BatmanAoD pulled up some benchmarks showing that, for some datasets, `BTreeMap` is actually faster. This may be an interesting thing for @Shadow53 to benchmark, and if it was indeed faster for our use case, it would just so happen to solve this problem for us.

Reply:
I've reviewed the pyQuil tests and a/b tested them with and without having the hashable instructions in question and confirmed they aren't needed. I've gone ahead and removed the Hash implementation for those instructions.